### PR TITLE
[browser] new EmccEnableAssertions and EmccEnvironment MSBuild props

### LINKED ISSUE #62708: [wasm] Allow user to set `-s ENVIRONMENT` setting
The [environment setting](https://github.com/emscripten-core/emscripten/blob/2.0.34/src/settings.js#L616-L641) can save small amount of size in `dotnet.js`.

It would be nice to autodetect which environment is needed during build or at least have a way to override it.

### LINKED ISSUE #74346: [wasm] make it possible to enable emscripten assertions `-s ASSERTIONS=1`
- [x] migrate Blazor to new asset loading with `maloc` after `onRuntimeInitialized`
- [ ] `-s ASSERTIONS=1` for debug build of runtime. It emits asserts to C and JS, it helps with debugging.

See also 
- https://github.com/dotnet/aspnetcore/issues/13915
- https://github.com/dotnet/runtime/pull/62254

### PR DESCRIPTION
**Pull request by pavelsavara**
- added `EmccEnableAssertions` boolean
    -  makes it easier to debug issues with wasm/native code and posix emulation
    - removed `_EmccAssertionLevelDefault`
- added `EmccEnvironment` comma separated list. Default `web,webview,worker,node,shell`.
    - allows to build slightly smaller dotnet.js without code for nodeJS
- removes `Module.FS_readFile`
- makes `dotnet.js.symbols` into non-VFS asset. Now it's loaded before mono starts.
- added `EmccExportedLibraryFunction` for `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE`

Usage:
```
    <EmccAssertions Condition="'$(Configuration)' == 'Debug'">true</EmccAssertions>
    <EmccEnvironment>web</EmccEnvironment>
```

Fixes https://github.com/dotnet/runtime/issues/62708
Fixes https://github.com/dotnet/runtime/issues/74346
Related https://github.com/dotnet/aspnetcore/issues/13915

Tested with latest Blazor locally 

image

### CODE REVIEW

File: src/mono/wasm/wasm.proj [hunk 1]
```diff
@@ -196,7 +196,6 @@
       <EmccExportedRuntimeMethod Include="FS_createDataFile" />
       <EmccExportedRuntimeMethod Include="removeRunDependency" />
       <EmccExportedRuntimeMethod Include="addRunDependency" />
-      <EmccExportedRuntimeMethod Include="FS_readFile" />
```
Comment (lewing) [inline src/mono/wasm/wasm.proj:199]: Is this not needed?

Comment (pavelsavara) [reply src/mono/wasm/wasm.proj:199]: emcc complains it's invalid when I set asserts on. 
It's used for symbolicator on JS side.
https://github.com/dotnet/runtime/blob/f604caf8d4014336bb04359cdfed70767f2975ae/src/mono/wasm/runtime/logging.ts#L190

Why is that in VFS @radical ?

Comment (radical) [reply src/mono/wasm/wasm.proj:199]: As long as this can read it, sure.

Comment (radical) [reply src/mono/wasm/wasm.proj:199]: Will it mean not early enough?

Comment (pavelsavara) [reply src/mono/wasm/wasm.proj:199]: VFS is like any other asset. If we need it before emscripten start (before wasm) we need some other approach.

Comment (radical) [reply src/mono/wasm/wasm.proj:199]: No, not earlier than what it is now.

Comment (pavelsavara) [reply src/mono/wasm/wasm.proj:199]: Right now we are not blocking emscripten start till after all VFS arrived.
We are blocking mono till after all assets arrived.

Comment (pavelsavara) [reply src/mono/wasm/wasm.proj:200]: done

File: src/mono/wasm/runtime/types.ts [hunk 2]
```diff
@@ -201,6 +201,7 @@ export type AssetBehaviours =
     | "vfs" // load asset into the virtual filesystem (for fopen, File.Open, etc)
     | "dotnetwasm" // the binary of the dotnet runtime
     | "js-module-threads" // the javascript module for threads
+    | "symbols" // the javascript module for threads
```
Comment (kg) [inline src/mono/wasm/runtime/types.ts:204]: incorrect comment

File: src/mono/wasm/wasm.proj [hunk 3]
```diff
@@ -197,14 +197,13 @@
...
       <EmccExportedRuntimeMethod Include="addFunction" />
-      <EmccExportedRuntimeMethod Include="getWasmTableEntry" />
 
       <EmccExportedFunction Include="_malloc" />
       <EmccExportedFunction Include="_memalign" />
       <EmccExportedFunction Include="_htons" />
       <EmccExportedFunction Include="_ntohs" />
+      <EmccExportedFunction Include="getWasmTableEntry" />
```
Comment (kg) [inline src/mono/wasm/wasm.proj:206]: Odd, it's an ExportedFunction not an ExportedRuntimeMethod?

File: src/mono/wasm/wasm.proj [hunk 4]
```diff
@@ -331,7 +330,7 @@
       <CMakeConfigurationEmccFlags Condition="'$(Configuration)' == 'Debug'">-g -Os -s -DDEBUG=1 -DENABLE_AOT_PROFILER=1 -DENABLE_BROWSER_PROFILER=1</CMakeConfigurationEmccFlags>
       <CMakeConfigurationEmccFlags Condition="'$(Configuration)' == 'Release'">-Oz -DENABLE_BROWSER_PROFILER=1</CMakeConfigurationEmccFlags>
 
-      <CMakeConfigurationLinkFlags Condition="'$(Configuration)' == 'Debug'"  >$(CMakeConfigurationEmccFlags)</CMakeConfigurationLinkFlags>
+      <CMakeConfigurationLinkFlags Condition="'$(Configuration)' == 'Debug'"  >$(CMakeConfigurationEmccFlags) -s ASSERTIONS=1 </CMakeConfigurationLinkFlags>
```
Comment (kg) [inline src/mono/wasm/wasm.proj:333]: I'm ok with 'assertions on in debug' but I don't think we should do it this way, we should be doing the equivalent of setting EmccAssertions

File: src/tasks/WasmAppBuilder/WasmAppBuilder.cs [hunk 5]
```diff
@@ -199,6 +204,9 @@ protected override bool ExecuteInternal()
             {
                 config.Assets.Add(new ThreadsWorkerEntry (name, Utils.ComputeIntegrity(item.ItemSpec)));
             }
+            else if(name == "dotnet.js.symbols"){
```
Comment (kg) [inline src/tasks/WasmAppBuilder/WasmAppBuilder.cs:207]: whitespace

**Review by kg — APPROVED** [review]
Looks good with the caveat that I'm not an expert on the debugging bits

File: src/mono/wasm/wasm.proj (hunk 4 above)
Comment (pavelsavara) [reply src/mono/wasm/wasm.proj:333]: This is never public facing, we are not giving out Debug binaries. That's why I did it this simplest possible way.

File: src/mono/wasm/wasm.proj (hunk 3 above)
Comment (pavelsavara) [reply src/mono/wasm/wasm.proj:206]: ok, right. Let me try again

https://github.com/emscripten-core/emscripten/blob/c4693feae0e366cf7741e9e1725fda5539605d4d/ChangeLog.md?plain=1#L337C7-L338

Comment (pavelsavara) [reply src/mono/wasm/wasm.proj:206]: We are on emscripten `3.1.12` and above note is in `3.1.16`.

Comment (pavelsavara) [reply src/mono/wasm/wasm.proj:206]: I guess I could wait with this PR until https://github.com/dotnet/runtime/pull/81215 is merged.

cc @radekdoulik

File: src/mono/wasm/build/WasmApp.targets [hunk 6]
```diff
@@ -49,6 +49,8 @@
       - $(EmccFlags)                        - Emcc flags used for both compiling native files, and linking
       - $(EmccExtraLDFlags)                 - Extra emcc flags for linking
       - $(EmccExtraCFlags)                  - Extra emcc flags for compiling native files
+      - $(EmccAssertions)                   - Corresponds to `ASSERTIONS` arg for emcc. Default false.
```
Comment (radical) [inline src/mono/wasm/build/WasmApp.targets:52]: If we want to support only `-sASSERTIONS=1` via `EmccAssertions: bool`, then it would be better to rename this to `EmccEnableAssertions`.

File: src/tasks/WasmAppBuilder/WasmAppBuilder.cs [hunk 7]
```diff
@@ -199,6 +204,10 @@ protected override bool ExecuteInternal()
             {
                 config.Assets.Add(new ThreadsWorkerEntry (name, Utils.ComputeIntegrity(item.ItemSpec)));
             }
+            else if(name == "dotnet.js.symbols")
+            {
+                config.Assets.Add(new SymbolsData (name, Utils.ComputeIntegrity(item.ItemSpec)));
```
Comment (radical) [inline src/tasks/WasmAppBuilder/WasmAppBuilder.cs:209]:
```suggestion
                config.Assets.Add(new SymbolsData(name, Utils.ComputeIntegrity(item.ItemSpec)));
```

File: src/mono/wasm/build/WasmApp.Native.targets [hunk 8]
```diff
@@ -419,7 +419,10 @@
     </PropertyGroup>
     <ItemGroup>
       <!-- order matters -->
+      <!-- some flags will be duplicated on the commandline and in the .rsp file. The last wins. -->
       <_EmccLDSFlags Include="-s INITIAL_MEMORY=$(EmccInitialHeapSize)" />
+      <_EmccLDSFlags Condition="'$(EmccEnvironment)' != ''" Include="-s ENVIRONMENT=&quot;$(EmccEnvironment)&quot;" />
```
Comment (maraf) [inline src/mono/wasm/build/WasmApp.Native.targets:424]: How does it work? Does it also drop *our* environment conditions?

File: src/mono/wasm/build/WasmApp.targets (hunk 6 above)
Comment (pavelsavara) [reply src/mono/wasm/build/WasmApp.targets:52]: I just noticed there is also `_EmccAssertionLevelDefault`

File: src/mono/wasm/runtime/assets.ts [hunk 9]
```diff
@@ -153,6 +156,12 @@ export async function mono_download_assets(): Promise<void> {
                         await beforeOnRuntimeInitialized.promise;
                         _instantiate_asset(asset, url, data);
                     }
+                    if (asset.behavior === "symbols") {
+                        await instantiate_symbols_asset(asset);
+                        asset.pendingDownloadInternal = null as any; // GC
+                        asset.pendingDownload = null as any; // GC
+                        asset.buffer = null as any; // GC
```
Comment (maraf) [inline src/mono/wasm/runtime/assets.ts:163]: Can we consolidate these with the same code in the if above?

File: src/mono/wasm/build/WasmApp.Native.targets [hunk 10]
```diff
@@ -449,6 +450,7 @@
       <_EmccLinkStepArgs Include="-o &quot;$(_WasmIntermediateOutputPath)dotnet.js&quot;" />
       <_WasmLinkDependencies Include="$(_EmccLinkRsp)" />
 
+      <_EmccLinkStepArgs Include="-s DEFAULT_LIBRARY_FUNCS_TO_INCLUDE=$(_EmccExportedLibraryFunction)" Condition="'@(_EmccExportedLibraryFunction)' != ''" />
```
Comment (maraf) [inline src/mono/wasm/build/WasmApp.Native.targets:453]:
```suggestion
      <_EmccLinkStepArgs Include="-s DEFAULT_LIBRARY_FUNCS_TO_INCLUDE=$(_EmccExportedLibraryFunction)" Condition="'$(_EmccExportedLibraryFunction)' != ''" />
```

File: src/mono/wasm/wasm.proj [hunk 11]
```diff
@@ -272,6 +270,7 @@
       <EmccExportedFunction Include="_free" />
     </ItemGroup>
     <PropertyGroup>
+      <_EmccExportedLibraryFunction>&quot;[@(EmccExportedLibraryFunction -> '%27%(Identity)%27', ',')]&quot;</_EmccExportedLibraryFunction>
```
Comment (maraf) [inline src/mono/wasm/wasm.proj:273]: Add comment about `EmccExportedLibraryFunction` in the https://github.com/dotnet/runtime/blob/6706d3f04142d4a635c1266c62e4219336f4ea1e/src/mono/wasm/build/WasmApp.targets#L80 and the https://github.com/dotnet/runtime/blob/6706d3f04142d4a635c1266c62e4219336f4ea1e/src/mono/wasi/build/WasiApp.targets#L63

Comment (pavelsavara) [reply src/mono/wasm/wasm.proj:273]: it's already in `WasmApp.targets` and there is no JS in WASI

File: src/mono/wasm/runtime/assets.ts [hunk 12]
```diff
@@ -611,3 +596,10 @@ export async function wait_for_all_assets() {
 export function mono_wasm_get_loaded_files(): string[] {
     return runtimeHelpers.loadedFiles;
 }
+
+export function cleanupAsset(asset: AssetEntryInternal) {
+    // give GC change to collect resources
```
Comment (maraf) [inline src/mono/wasm/runtime/assets.ts:601]:
```suggestion
    // give GC chance to collect resources
```

**Review by maraf — APPROVED** [review]
Looks good to me 👍

**Comment by pavelsavara** [comment]
There was probably unrelated `DebuggerTests.BreakpointTests.StepThroughOrNonUserCodeAttributeWithUserBp` on last CI run.

**Comment by pavelsavara** [comment]
Failures are known 'NuGet-Migrations'
